### PR TITLE
use tempfile if cleanup=True on render

### PR DESCRIPTION
**Pull request by pyeguy**
this addresses #45 by checking if `cleanup=True` in the `render` method and if so calling `tempfile.mksfile('.gv')` when creating the `filepath` 

This works for `view(cleanup=True)` as well since this calls `render`

### CODE REVIEW

**Comment by pyeguy** [comment]
whoops, i should have run the tests before the PR. Not too sure how to work around this assertion... need to call `save` w/o the `filepath` of the temp file when `cleanup=True` 

I'll think about it.

```python
================================== FAILURES ===================================
_________________________________ test_render _________________________________

mocker = <pytest_mock.MockFixture object at 0x000001EB77090438>
render = <MagicMock name='render' id='2110826022392'>
source = <graphviz.files.Source object at 0x000001EB77078080>

    def test_render(mocker, render, source):
        save = mocker.patch.object(source, 'save')
        _view = mocker.patch.object(source, '_view')
        remove = mocker.patch('os.remove')

        assert source.render(cleanup=True, view=True) is render.return_value

>       save.assert_called_once_with(None, None)
E       AssertionError: Expected call: save(None, None)
E       Actual call: save('C:\\Users\\cpye\\AppData\\Local\\Temp\\tmpfh0s38xw.gv')
E
E       pytest introspection follows:
E
E       Args:
E       assert ('C:\\Users\\...fh0s38xw.gv',) == (None, None)
E         At index 0 diff: 'C:\\Users\\cpye\\AppData\\Local\\Temp\\tmpfh0s38xw.gv' != None
E         Right contains more items, first extra item: None
E         Use -v to get the full diff

tests\test_files.py:106: AssertionError
===================== 1 failed, 79 passed in 1.26 seconds =====================
```

**Comment by xflr6** [comment]
This breaks the following code (not backwards compatible):

```python
>>> import graphviz, os
>>> g = graphviz.Graph()
>>> g.node('spam')
>>> path = g.render(os.path.expanduser('~/spam.gv'), cleanup=True)
>>> assert path.endswith('spam.gv.pdf')
```

File: graphviz/files.py [hunk 1]
```diff
@@ -170,13 +171,14 @@ def render(self, filename=None, directory=None, view=False, cleanup=False):
...
-        filepath = self.save(filename, directory)
+        if cleanup:
+            filepath = self.save(tempfile.mkstemp('.gv')[1])
+        
+        else:
+            filepath = self.save(filename, directory)
 
         rendered = backend.render(self._engine, self._format, filepath)
 
-        if cleanup:
```
Comment (xflr6) [inline graphviz/files.py]: Why remove the actual cleanup here?

File: graphviz/files.py [hunk 2]
```diff
@@ -170,13 +171,14 @@ def render(self, filename=None, directory=None, view=False, cleanup=False):
             subprocess.CalledProcessError: If the exit status is non-zero.
             RuntimeError: If viewer opening is requested but not supported.
         """
-        filepath = self.save(filename, directory)
+        if cleanup:
+            filepath = self.save(tempfile.mkstemp('.gv')[1])
```
Comment (xflr6) [inline graphviz/files.py]: I think one would prefer to use the file object from `tempfile.NamedTemporaryFile(delete=False)` within `save()`, see also https://www.logilab.org/blogentry/17873

File: graphviz/files.py (hunk 1 above)
Comment (pyeguy) [inline graphviz/files.py]: fair enough, wasn't thinking of the other use cases.

File: graphviz/files.py (hunk 2 above)
Comment (pyeguy) [inline graphviz/files.py]: after looking at the code more this makes total sense. seems like it involves adding a `cleanup` arg to `save` though and modifying the test from `save.assert_called_once_with(None, None)` to save.assert_called_once_with(None, None, True)`

File: graphviz/files.py [hunk 3]
```diff
@@ -143,7 +144,14 @@ def save(self, filename=None, directory=None):
         if directory is not None:
             self.directory = directory
 
-        filepath = self.filepath
+        if self.filename is None and self.directory is None and cleanup:
```
Comment (xflr6) [inline graphviz/files.py]: `self.filename` is never `None` (see `__init__()`) so this branch never gets executed.
Still, I think using a `tempfile` should either be triggered by a dedicated flag or a special value for `filename` (i.e. it should be orthogonal to `cleanup`).

File: graphviz/files.py [hunk 4]
```diff
@@ -143,7 +144,14 @@ def save(self, filename=None, directory=None):
...
+            nmdtmpfile = tempfile.NamedTemporaryFile(suffix='.gv', delete=False)
+            nmdtmpfile.close()
```
Comment (xflr6) [inline graphviz/files.py]: The conentional use case for `NamedTemporaryFile` would be to use it directly as a file object (i.e. to use its `.write()`-method).

**Pull request closed without merging**